File: authEndpoints.js
/**
 * Authentication endpoints for MATRIYA RAG System
 */
import express from 'express';
import { User } from './database.js';
import {
  authenticateUser,
  createUser,
  verifyToken,
  getUserByUsername,
  getUserByEmail,
  createAccessToken,
  ACCESS_TOKEN_EXPIRE_MINUTES_EXPORT as ACCESS_TOKEN_EXPIRE_MINUTES
} from './auth.js';
import { getDb } from './database.js';
import logger from './logger.js';

const router = express.Router();

/**
 * Get current authenticated user from token
 */
async function getCurrentUser(req) {
  const authHeader = req.headers.authorization;
  if (!authHeader) {
    return null;
  }
  
  // Extract token from "Bearer <token>"
  try {
    const parts = authHeader.split(' ');
    if (parts.length !== 2 || parts[0].toLowerCase() !== 'bearer') {
      return null;
    }
    const token = parts[1];
    
    const payload = verifyToken(token);
    if (!payload) {
      return null;
    }
    
    const username = payload.sub;
    if (!username) {
      return null;
    }
    
    // Get user from database
    return await getUserByUsername(username);
  } catch (e) {
    return null;
  }
}

/**
 * Middleware to require authentication
 */
async function requireAuth(req, res, next) {
  const user = await getCurrentUser(req);
  if (!user) {
    return res.status(401).json({
      error: "Invalid authentication credentials"
    });
  }
  req.user = user;
  next();
}

/**
 * Create a new user account
 * 
 * JSON body:
 *   username: Username
 *   email: Email address
 *   password: Password
 *   full_name: Optional full name
 * 
 * Returns:
 *   Access token and user information
 */
router.post("/signup", async (req, res) => {
  try {
    const { username, email, password, full_name } = req.body;
    
    if (!username || !email || !password) {
      return res.status(400).json({ error: "username, email, and password are required" });
    }
    
    // Basic email validation
    if (!email.includes('@')) {
      return res.status(400).json({ error: "Invalid email format" });
    }
    
    // Check if username already exists
    if (await getUserByUsername(username)) {
      return res.status(400).json({ error: "Username already registered" });
    }
    
    // Check if email already exists
    if (await getUserByEmail(email)) {
      return res.status(400).json({ error: "Email already registered" });
    }
    
    // Create user
    const user = await createUser(username, email, password, full_name);
    
    // Create access token
    const accessToken = createAccessToken(
      { sub: user.username },
      ACCESS_TOKEN_EXPIRE_MINUTES
    );
    
    return res.json({
      access_token: accessToken,
      token_type: "bearer",
      user: {
        id: user.id,
        username: user.username,
        email: user.email,
        full_name: user.full_name,
        is_admin: user.is_admin
      }
    });
  } catch (e) {
    logger.error(`Signup error: ${e.message}`);
    return res.status(500).json({ error: `Signup failed: ${e.message}` });
  }
});

/**
 * Login and get access token
 * 
 * JSON body:
 *   username: Username
 *   password: Password
 * 
 * Returns:
 *   Access token and user information
 */
router.post("/login", async (req, res) => {
  try {
    const { username, password } = req.body;
    
    if (!username || !password) {
      return res.status(400).json({ error: "username and password are required" });
    }
    
    const user = await authenticateUser(username, password);
    if (!user) {
      return res.status(401).json({ error: "Incorrect username or password" });
    }
    
    // Update last login
    try {
      user.last_login = new Date();
      await user.save();
    } catch (e) {
      // Don't fail login if last_login update fails
      logger.warn(`Failed to update last_login: ${e.message}`);
    }
    
    // Create access token
    const accessToken = createAccessToken(
      { sub: user.username },
      ACCESS_TOKEN_EXPIRE_MINUTES
    );
    
    return res.json({
      access_token: accessToken,
      token_type: "bearer",
      user: {
        id: user.id,
        username: user.username,
        email: user.email,
        full_name: user.full_name,
        is_admin: user.is_admin
      }
    });
  } catch (e) {
    logger.error(`Login error: ${e.message}`);
    return res.status(500).json({ error: `Login failed: ${e.message}` });
  }
});

/**
 * Get current user information
 * 
 * Returns:
 *   Current user information
 */
router.get("/me", requireAuth, async (req, res) => {
  const user = req.user;
  return res.json({
    id: user.id,
    username: user.username,
    email: user.email,
    full_name: user.full_name,
    is_admin: user.is_admin,
    created_at: user.created_at ? user.created_at.toISOString() : null
  });
});

export { router as authRouter, getCurrentUser, requireAuth };

File: auth_endpoints.py
"""
Authentication endpoints for MATRIYA RAG System
"""
from fastapi import APIRouter, Depends, HTTPException, status
from sqlalchemy.orm import Session
from pydantic import BaseModel, EmailStr
from typing import Optional
from datetime import timedelta
from database import get_db
from auth import (
    authenticate_user, create_user, verify_token,
    get_user_by_username, get_user_by_email,
    create_access_token, ACCESS_TOKEN_EXPIRE_MINUTES
)
from fastapi.security import HTTPBearer, HTTPAuthorizationCredentials
from database import User

router = APIRouter(prefix="/auth", tags=["authentication"])
security = HTTPBearer()


class UserSignup(BaseModel):
    username: str
    email: EmailStr
    password: str
    full_name: Optional[str] = None


class UserLogin(BaseModel):
    username: str
    password: str


class TokenResponse(BaseModel):
    access_token: str
    token_type: str
    user: dict


class UserResponse(BaseModel):
    id: int
    username: str
    email: str
    full_name: Optional[str]
    is_admin: bool
    created_at: str


def get_current_user(
    credentials: HTTPAuthorizationCredentials = Depends(security),
    db: Session = Depends(get_db)
):
    """Get current authenticated user"""
    token = credentials.credentials
    payload = verify_token(token)
    if payload is None:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid authentication credentials",
            headers={"WWW-Authenticate": "Bearer"},
        )
    username: str = payload.get("sub")
    if username is None:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="Invalid authentication credentials",
            headers={"WWW-Authenticate": "Bearer"},
        )
    # Optimized query - direct query for faster response
    user = db.query(User).filter(User.username == username).first()
    if user is None:
        raise HTTPException(
            status_code=status.HTTP_401_UNAUTHORIZED,
            detail="User not found",
            headers={"WWW-Authenticate": "Bearer"},
        )
    return user


@router.post("/signup", response_model=TokenResponse)
async def signup(user_data: UserSignup, db: Session = Depends(get_db)):
    """
    Create a new user account
    
    Args:
        user_data: User signup information
        
    Returns:
        Access token and user information
    """
    # Check if username already exists
    if get_user_by_username(db, user_data.username):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Username already registered"
        )
    
    # Check if email already exists
    if get_user_by_email(db, user_data.email):
        raise HTTPException(
            status_code=status.HTTP_400_BAD_REQUEST,
            detail="Email already registered"
        )
    
    # Create user
    user = create_user(
        db=db,
        username=user_data.username,
        email=user_data.email,
        password=user_data.password,
        full_name=user_data.full_name
    )
    
    # Create access token
    access_token = create_access_token(
        data={"sub": user.username},
        expires_delta=timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
    )
    
    return {
        "access_token": access_token,
        "token_type": "bearer",
        "user": {
            "id": user.id,
            "username": user.username,
            "email": user.email,
            "full_name": user.full_name,
            "is_admin": user.is_admin
        }
    }


@router.post("/login", response_model=TokenResponse)
async def login(login_data: UserLogin, db: Session = Depends(get_db)):
    """
    Login and get access token
    
    Args:
        login_data: Username and password
        
    Returns:
        Access token and user information
    """
    try:
        user = authenticate_user(db, login_data.username, login_data.password)
        if not user:
            raise HTTPException(
                status_code=status.HTTP_401_UNAUTHORIZED,
                detail="Incorrect username or password",
                headers={"WWW-Authenticate": "Bearer"},
            )
        
        # Update last login (async - don't block response)
        from datetime import datetime
        try:
            user.last_login = datetime.utcnow()
            db.commit()
        except:
            db.rollback()
            # Don't fail login if last_login update fails
        
        # Create access token
        access_token = create_access_token(
            data={"sub": user.username},
            expires_delta=timedelta(minutes=ACCESS_TOKEN_EXPIRE_MINUTES)
        )
        
        return {
            "access_token": access_token,
            "token_type": "bearer",
            "user": {
                "id": user.id,
                "username": user.username,
                "email": user.email,
                "full_name": user.full_name,
                "is_admin": user.is_admin
            }
        }
    except HTTPException:
        raise
    except Exception as e:
        logger.error(f"Login error: {e}")
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail=f"Login failed: {str(e)}"
        )


@router.get("/me", response_model=UserResponse)
async def get_current_user_info(current_user: User = Depends(get_current_user)):
    """
    Get current user information
    
    Returns:
        Current user information
    """
    return {
        "id": current_user.id,
        "username": current_user.username,
        "email": current_user.email,
        "full_name": current_user.full_name,
        "is_admin": current_user.is_admin,
        "created_at": current_user.created_at.isoformat() if current_user.created_at else None
    }
